module/Decision/src/Decision/Controller/FileBrowser/LocalFileReader.php

```diff
@@ -0,0 +1,83 @@
+<?php
+
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+
+namespace Decision\Controller\FileBrowser;
+
+/**
+ * Description of LocalFileReader
+ *
+ * @author s134399
+ */
+class LocalFileReader implements FileReader
+{
+
+    /**
+     * The location in the local filesystem that is considered the 'root' for this browser
+     * @var string
+     */
+    static private $root;
+
+    public function __construct($root)
+    {
+        $this->root = $root;
+    }
+
+    public function downloadFile($path)
+    {
+        $fullPath = $this->root . $path;
+        if (!is_file($fullPath)) {
+            return null;
+        }
+        $response = new \Zend\Http\Response\Stream();
+        $response->setStream(fopen('file://' . $fullPath, 'r'));
+        $response->setStatusCode(200);
+        $headers = new \Zend\Http\Headers();
+        $headers->addHeaderLine('Content-Type', 'application/octet-stream')
+                ->addHeaderLine('Content-Disposition', 'attachment; filename="' . end(explode('/', $fullPath)) . '"')
+                ->addHeaderLine('Content-Length', filesize($fullPath));
+        $response->setHeaders($headers);
+        return $response;
+    }
+
+    public function listDir($path)
+    {
+        //remove the trailing slash from the dir
+        $fullPath = $this->root . $path;
+        if (!is_dir($fullPath)) {
+            return null;
+        }
+        //We can insert an additional /, except when when $path is the root
+        $delimiter = $path !== '' ? '/': '';
+        $dircontents = scandir($fullPath);
+        $files = [];
+        foreach ($dircontents as $dircontent) {
+            if ($dircontent[0]==='.') {
+                continue;
+            }
+            if (is_dir($fullPath . '/' . $dircontent)) {
+                $kind = 'dir';
+            } elseif (is_file($fullPath . '/' . $dircontent)) {
+                $kind = 'file';
+            } else {
+                //Ignore all strange filesystem thingies like symlinks and such
+                continue;
+            }
+            $files[] = new FileNode(
+                $kind,
+                htmlspecialchars($path . $delimiter . $dircontent),
```

Comment:
Wouldn't it be more 'correct' to store the valid paths in the FileNode object? This way, the nodes still contain the actual valid paths, like the comments added in the class state and you can make the call to htmlspecialchars in the views.
Otherwise I would suggest making a small change in the documentation of the FileNode class.

Reply:
I'd prefer to invoke htmlspecialchars sooner rather than later. If there are truly filepaths that have html chars in them, how on earth would I be able to access them? In other words, I'm working on the reasonable assumption that given that a path is valid, htmlspecialchars does nothing.

I think the documentation in FileNode is correct. It represents a node in a 'filesystem', where 'filesystem' here means 'whatever it is FileReader browses through'. The 'path' is valid in the sense that it is the location of the node as far as the FileReader is concerned. (hmtlspecialchars doesn't change this, as I already mentioned)